### PR TITLE
feat: support SOCKS proxy

### PR DESCRIPTION
**Pull request by jirimoravcik**
This PR introduces support for SOCKS protocol.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 [![npm version](https://badge.fury.io/js/proxy-chain.svg)](http://badge.fury.io/js/proxy-chain)
 
-A programmable proxy server (think Squid) with support for SSL/TLS, authentication, upstream proxy chaining,
+A programmable proxy server (think Squid) with support for SSL/TLS, authentication, upstream proxy chaining (supports SOCKS protocol),
```
Comment (jancurn) [inline README.md:5]: Which versions of SOCKS?

File: src/server.ts [hunk 2]
```diff
@@ -18,6 +18,10 @@ import { Socket } from './socket';
 import { normalizeUrlPort } from './utils/normalize_url_port';
 import { badGatewayStatusCodes } from './statuses';
 import { customConnect } from './custom_connect';
+import { forwardSocks } from './forward_socks';
+import { chainSocks } from './chain_socks';
+
+const SOCKS_PROTOCOLS = ['socks:', 'socks4:', 'socks4a:', 'socks5:', 'socks5h:'];
```
Comment (jancurn) [inline src/server.ts:24]: Please can we document this list of supported protocols in README?

**Review by jancurn — APPROVED** [review]
Looks great, thanks Jirka! It would be great to add at least one unit/integration test for the SOCKS proxy, perhaps extend our existing test suite

**Comment by dumbasPL** [comment]
might wanna remove `The SOCKS protocol is not supported yet` from the readme ;)

**Comment by jirimoravcik** [comment]
> might wanna remove `The SOCKS protocol is not supported yet` from the readme ;)

Will do, thanks for noticing 🙏

**Comment by jirimoravcik** [comment]
> `anonymizeProxy` is still throwing the error, can this be removed?
> 
> https://github.com/apify/proxy-chain/blob/a1ac3f287e895270f4e34306ac754a62a9848fa4/src/anonymize_proxy.ts#L41-L45

Thanks for reporting this, fix is in https://github.com/apify/proxy-chain/pull/545 and will be merged and released soon.

**Pull request merged**